### ISSUE TITLE
[AppBar] Elements below the App Bar can be read with voice over

### ISSUE THREAD

**Issue by gambard** [material-components/material-components-ios#2853]
### Overview
When a collection is scrolled below the header, it is possible to read the content of the cells below it with voice over.

### Reproduction steps
 - Open a collection example on the catalog (e.g. Cell Editing Example)
 - Scroll it such as some cells are below the header
 - Activate voice over
 - Touch between the "back" arrow and the title of the header

What is the expected result?
 - You shouldn't be able to select the cells below the title.

What happens instead?
 - You should be able to select them.

### Version number
 - 44

---
#### Internal data
- Associated internal bug: [b/117179244](http://b/117179244)

**Comment by jverkoey** [comment]
I dug into this yesterday and learned the following.

1. In general, VoiceOver is able to select accessibility elements that are obscured by other views. This can be reproduced by creating a label, placing it behind a fully-opaque view, and attempting to tap the label.
2. UINavigationBar, however, deviates from this behavior. I do not know how, because the accessibility frames for all of the elements are much smaller than the selectable areas (the region within which a tap will result in one of the elements being focused by voice over). As best I can tell, UIINavigationBar is somehow able to influence how VoiceOver does selection when taps occur within the navigation bar's frame. For example, if you tap the space between the title and the buttons on a navigation bar, the closer of either the title or the buttons will be selected. This is not standard voice over selection behavior, and I do not know how UINavigationBar implements this behavior.

The only solution that I can think of, which is less than ideal, is for our navigation bar to adjust the accessibility frames such that they completely cover the navigation bar's frame, simulating the behavior of UINavigationBar. This will mean that our accessibility frames will appear fairly large overall, but I think that this will allow us to achieve the desired voice over behavior influence described in this issue.

**Comment by galiak11** [comment]
It looks like the winning solution for this is to add a UIAccessibilityElement as a subview of the UINavigationBar with an accessibilityFrame that is the size of the navigation bar, and return it as the last element of accessibilityElements array.
This solves the issue in that elements underneath the navigation bar are no longer "seen" by voice over, however, it adds an additional accessible element when swiping right/left to iterate through the accessibility elements. Not clear what the label should be for that extra element. Should it read something like "Navigation Bar"?
Also, HitTest doesn't seem to work well with voice over. Elements beneath the navigation bar still seem to be read, both on single and double taps.

**Comment by galiak11** [comment]
Also note this [appledev-discuss question](https://groups.google.com/a/google.com/forum/#!topic/appledev-discuss/280G4Bf1e0Y).

**Comment by jverkoey** [comment]
@galiak11 How feasible do you feel this bug is to resolve? If we don't feel like there's a good path forward, let's mark as "Working as intended" and close the issue.

**Comment by jverkoey** [comment]
Un-assigning due to inactivity. Please re-assign if you are working on this.

**Comment by randallli** [comment]
Works as intended.

**Comment by romoore** [comment]
Randall - can you clarify why this issue has suddenly changed to "works as intended"? My understanding is that VoiceOver should not be accessing elements not visible on-screen. Just as users that do not use VoiceOver cannot access them, there's no reason why VoiceOver should be accessing them.

**Comment by romoore** [comment]
@jverkoey The proposed solution to use custom UIAccessibilityElement should not be particularly difficult. One of those could be the custom title view if set, the others would be the buttons.  This is the path taken by many UIKit controls and is well-supported in UIKit.

**Comment by wenyuzhang666** [comment]
The cell underneath the header could be accessible in some cases. For example, when header is transparent, the cell underneath the header would be visible to users visually impaired. It would be tricky on our side to make things underneath the header invisible to accessibility.

On the other side, if the header is transparent, users do not use VoiceOver are able to read the text on cell (though no tappable), the users who use VoiceOver should be able to access and read the text on cell as well.

Closing this issue after discussion. Please feel free to re-open if it becomes a critical accessibility blocker.
